File: code/763.partition-labels.java
/*
 * @lc app=leetcode id=763 lang=java
 *
 * [763] Partition Labels
 */

// @lc code=start

import java.util.HashMap;
import java.util.Map;

class Solution {
    public List<Integer> partitionLabels(String s) {
        Map<Character, Integer> map = new HashMap<>();

        for (int i = 0; i < s.length(); i++) {
            map.put(s.charAt(i), i);
        }

        List<Integer> res = new ArrayList<>();

        int start = 0;
        int end = 0;
        for (int i = 0; i < s.length(); i++) {;
            end = Math.max(end, map.get(s.charAt(i)));
            if (i == end) {
                res.add(end - start + 1);
                start = end + 1;
            }
        }
        return res;
    }
}
// @lc code=end

File: code/763.partition-labels.go
/*
 * @lc app=leetcode id=763 lang=golang
 *
 * [763] Partition Labels
 */

// @lc code=start
func partitionLabels(s string) []int {
	var marks [26]int

	for i := 0; i < len(s); i++ {
		marks[s[i]-'a'] = i
	}

	res := make([]int, 0)
	start, end := 0, 0

	for i := 0; i < len(s); i++ {
		end = max(end, marks[s[i]-'a'])
		if i == end {
			res = append(res, end-start+1)
			start = end + 1
		}
	}
	return res
}

func max(a, b int) int {
	if a < b {
		a = b
	}
	return a
}

// @lc code=end

